source/opt/licm_pass.cpp

```diff
@@ -0,0 +1,138 @@
+    ir_context = context;
+    modified = ProcessIRContext();
+  }
+
+  return modified ? Status::SuccessWithChange : Status::SuccessWithoutChange;
+}
+
+bool LICMPass::ProcessIRContext() {
+  bool modified = false;
+  ir::Module* module = ir_context->module();
+
+  // Process each function in the module
+  for (ir::Function& f : *module) {
+    modified |= ProcessFunction(&f);
+  }
+  return modified;
+}
+
+bool LICMPass::ProcessFunction(ir::Function* f) {
+  bool modified = false;
+  ir::LoopDescriptor loop_descriptor{f};
+
+  // Process each loop in the function
+  for (ir::Loop& loop : loop_descriptor) {
+    modified |= ProcessLoop(&loop, f);
+  }
+  return modified;
+}
+
+bool LICMPass::ProcessLoop(ir::Loop* loop, ir::Function* f) {
+  // Process all nested loops first
+  for (ir::Loop* nested_loop : *loop) {
+    ProcessLoop(nested_loop, f);
+  }
+
+  std::queue<ir::Instruction*> loop_iv_instr{};
+  GatherAllImmediatelyInvariantInstructions(loop, &loop_iv_instr);
+
+  return ProcessInstructionList(loop, &loop_iv_instr);
+}
+
+void LICMPass::GatherAllImmediatelyInvariantInstructions(
+    ir::Loop* loop, std::queue<ir::Instruction*>* loop_iv_instr) {
+  // For each instruction in each basic block in the loop, check if it is
+  // invariant. If so, push it to our invariants queue.
+  for (uint32_t bb_id : loop->GetBlocks()) {
+    ir::BasicBlock* bb = ir_context->get_instr_block(bb_id);
+    for (ir::Instruction& inst : *bb) {
+      if (inst.opcode() != SpvOpPhi && !inst.HasSideEffects() &&
+          AllOperandsOutsideLoop(loop, &inst)) {
+        loop_iv_instr->push(&inst);
+      }
+    }
+  }
+}
+
+void LICMPass::HoistInstruction(ir::BasicBlock* pre_header_bb,
+                                ir::Instruction* inst) {
+  inst->InsertBefore(std::move(&(*pre_header_bb->tail())));
+  ir_context->set_instr_block(inst, pre_header_bb);
+}
+
+bool LICMPass::AllOperandsOutsideLoop(ir::Loop* loop, ir::Instruction* inst) {
+  analysis::DefUseManager* def_use_mgr = ir_context->get_def_use_mgr();
+  bool all_outside_loop = true;
+
+  const std::function<void(uint32_t*)> operand_test_lambda =
+      [&def_use_mgr, &loop, &all_outside_loop](uint32_t* id) {
+        if (loop->IsInsideLoop(def_use_mgr->GetDef(*id))) {
+          all_outside_loop = false;
+        }
+      };
+
+  inst->ForEachInId(operand_test_lambda);
+  return all_outside_loop;
+}
+
+bool LICMPass::ProcessInstructionList(
+    ir::Loop* loop, std::queue<ir::Instruction*>* loop_iv_instr) {
+  if (loop_iv_instr->empty()) {
+    return false;
+  }
+
+  ir::BasicBlock* pre_header_bb = loop->GetPreHeaderBlock();
+
+  // If the input instruction is invariant, push it onto the invariants queue
+  const std::function<void(ir::Instruction*)> check_users_now_invariant =
+      [this, &loop, &loop_iv_instr](ir::Instruction* inst) {
+        if (inst->opcode() != SpvOpPhi && !inst->HasSideEffects() &&
+            AllOperandsOutsideLoop(loop, inst)) {
+          loop_iv_instr->push(inst);
```

Comment:
What about instructions that cannot be moved?

Reply:
In fact, a barrier instruction for instance would prevent moving instructions downstream from it, even if they're LI.

Reply:
Should the pass ignore any loops containing barriers then?

Reply:
Why would a barrier stop us from moving things?  A barrier is more about memory.  If the instruction is not a load or store of some type, then we should be able to move it.  On that note, I did not really pay attention to what happens to loads.  They are considered combinators.  Even if all of its operands are defined outside of the loop, the same variable is stored to then we have to leave it.

Reply:
Yes, I meant instructions accessing shared resources.  So loads/stores and anything else that may have hidden shared inputs/outputs.